core/payment/src/api/provider.rs

```diff
@@ -326,8 +328,50 @@ async fn cancel_invoice(
     db: Data<DbExecutor>,
     path: Path<InvoiceId>,
     query: Query<Timeout>,
+    id: Identity,
 ) -> HttpResponse {
-    response::not_implemented() // TODO
+    let invoice_id = path.invoice_id.clone();
+    let node_id = id.identity;
+    let dao: InvoiceDao = db.as_dao();
+    let invoice = match dao.get(invoice_id.clone(), node_id).await {
+        Ok(Some(invoice)) => invoice,
+        Ok(None) => return response::not_found(),
+        Err(e) => return response::server_error(&e),
+    };
+
+    match invoice.status {
+        DocumentStatus::Issued => (),
+        DocumentStatus::Received => (),
+        DocumentStatus::Rejected => (),
+        DocumentStatus::Cancelled => return response::ok(Null),
+        DocumentStatus::Accepted | DocumentStatus::Settled | DocumentStatus::Failed => {
```

Comment:
should failed get its own response error message?

Reply:
If the payment failed, invoice must had been accepted beforehand. So the message is technically correct.